Add support for Scala 3

tgodzik opened the pull request:
Previously, we would need to compile the Scala compiler bridge in order to use it with a particular Scala version. Now, it's possible to download the jar, which is released together with a particular Scala version.

From other changes, a new compiler is needed together with Scala library, however it still reuses the Scala 2.13 library underneath.

Sample project this works with:

```kotlin
plugins {
    scala
}

repositories {
    mavenCentral()
    mavenLocal()
}

dependencies {
    implementation("org.scala-lang:scala3-library_3:3.0.1")
}
```

Fixes https://github.com/gradle/gradle/issues/16527


### Context
Scala 3 has been released in May and a lot of its' potential users use Gradle, which is de facto standard in the JVM world. This is why 

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

smarter commented on subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java:
```diff
@@ -113,43 +128,57 @@ static ZincScalaCompiler getCompiler(CacheRepository cacheRepository, HashedClas
-
         File libraryJar = findFile(ArtifactInfo.ScalaLibraryID, scalaClasspath);
-        File compilerJar = findFile(ArtifactInfo.ScalaCompilerID, scalaClasspath);
+        URL[] libraryUrls = new URL[]{libraryJar.toURI().toURL()};
```
This should contain both scala-library and scala3-library.

tgodzik replied:
Didn't seem to break anything, but sure I can add it!

tgodzik replied:
Fixed!

dubinsky commented:
@tgodzik Wow! Thank you so much!!!
My attempt (#18003) ran into 'java.util.ServiceConfigurationError: xsbti.compile.CompilerInterface2: dotty.tools.xsbt.CompilerBridge not a subtype'; your change does some classpath manipulations that probably work around that issue - thanks!

I that that to gain Scala 3 support, we have to use post-1.5 Zinc (that is why all the Zinc changes in my draft)...

tgodzik commented:
> @tgodzik Wow! Thank you so much!!!
> My attempt (#18003) ran into 'java.util.ServiceConfigurationError: xsbti.compile.CompilerInterface2: dotty.tools.xsbt.CompilerBridge not a subtype'; your change does some classpath manipulations that probably work around that issue - thanks!
> 
> I thought that to gain Scala 3 support, we have to use post-1.5 Zinc (that is why all the Zinc changes in my draft)...

We don't actually need that, there is a fallback step currently, which we will want to remove, but not necessarily now. For sure we should update Zinc as the next step. I didn't want to everything in one PR, but I will most likely follow up with that.

dubinsky commented:
> @dubinsky I see that you actually updated zinc, so maybe we could rebase your PR later on and update to latest zinc this way?

Sure, I am ready to help in any way. I never used Zinc before, though, so I do not really know what I am doing.
If you want to just use parts of my draft outright - it is fine with me :)

> I see that you also spotted some places that might require some investigation (commented with ` // TODO: support Scala3`). I think I need to take a look at `ScalaBasePlugin.java` and fix it here.

Looking forward to it!

By the way, does `IdeScalaConfigurer` need to be made Scala3-aware?
In my draft I centralized the library-finding logic so it could be [re-used](https://github.com/gradle/gradle/pull/18003/files#diff-132461705ecca69760d2b6e14c698493ea089443aaf9eb3aa42260fa61d84d6aR141) without code duplication, but as long as Scala 3 is supported by Gradle both on the command line and from IntelliJ Idea, I do not really care how it is done ;)

Thanks for all your work!

luis- commented:
> @tgodzik Wow! Thank you so much!!!
> 
> My attempt (#18003) ran into 'java.util.ServiceConfigurationError: xsbti.compile.CompilerInterface2: dotty.tools.xsbt.CompilerBridge not a subtype'; your change does some classpath manipulations that probably work around that issue - thanks!
> 
> 
> 
> I thought that to gain Scala 3 support, we have to use post-1.5 Zinc (that is why all the Zinc changes in my draft)...
> 
>  

I also thought zinc 1.5 was needed. First time digging into gradle code but over the weekend I managed to swap in all the VirtualFile to uptake zinc 1.5 to add scala 3 support. Thought I was close until I ran into xsbti.* errors and then ended up butchering my working tree trying to figure it out. From looking at this and #18003 its nice to see I wasn't completely off base with my attempt. Good learning experience for me.

GavinRay97 commented:
Woohoo! -- huge thank-you to @tgodzik, @dubinsky and @smarter!!

🙏 🎉

tgodzik commented:
The main issue wasn't actually zinc version, but wrong jars in the ScalaInstance. We can still upgrade it for sure, but it might be easier to do it in a separate pr

tgodzik commented:
> By the way, does `IdeScalaConfigurer` need to be made Scala3-aware?

Good question, I will check!

> In my draft I centralized the library-finding logic so it could be [re-used](https://github.com/gradle/gradle/pull/18003/files#diff-132461705ecca69760d2b6e14c698493ea089443aaf9eb3aa42260fa61d84d6aR141) without code duplication, but as long as Scala 3 is supported by Gradle both on the command line and from IntelliJ Idea, I do not really care how it is done ;)

 It's probably better to rework it but I wanted to keep it to a minimum, so that it is easier to review. I think We can also do a follow up.

tgodzik commented:
`IdeScalaConfigurer` should work correctly, I added some testing and also tried use it with Intellij, which seems to work alright.

@jastice are you able to confirm that's enough to make Scala 3 work for Gradle with Intellij? Or is there something more we need to do?

kamildoleglo commented:
> Also, anyone knows how I can get the PR reviewed?

Maybe @eskatos could help

dubinsky commented:
@tgodzik some observations when using your branch:
- compiler error messages sometimes give position of the offending code as **character** numbers (e.g., <300..312>), not line number and character in the line; I am not sure if this is the result of using version of Zinc that is not precise enough for Scala 3, or if some compiler option affects this, or if this new normal - but Idea does not seem equipped to handle this.
- if I configure Idea to build the project directly instead of using Gradle (File | Settings | Build, Execution, Deployment | Build Tools | Gradle | Build and run using: IntelliJ IDEA), I get:
```
scalac: Error: xsbti/compile/CompilerInterface2
java.lang.NoClassDefFoundError: xsbti/compile/CompilerInterface2
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:550)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:458)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:452)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:451)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at sbt.internal.inc.classpath.DualLoader.loadClass(DualLoader.scala:65)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:576)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1210)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1221)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator$1.run(ServiceLoader.java:1268)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator$1.run(ServiceLoader.java:1267)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1270)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1300)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1385)
	at sbt.internal.inc.AnalyzingCompiler.loadService(AnalyzingCompiler.scala:302)
	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:87)
	at org.jetbrains.jps.incremental.scala.local.IdeaIncrementalCompiler.compile(IdeaIncrementalCompiler.scala:57)
	at org.jetbrains.jps.incremental.scala.local.LocalServer.compile(LocalServer.scala:43)
	at org.jetbrains.jps.incremental.scala.remote.Main$.compileLogic(Main.scala:158)
	at org.jetbrains.jps.incremental.scala.remote.Main$.$anonfun$handleCommand$1(Main.scala:141)
	at org.jetbrains.jps.incremental.scala.remote.Main$.decorated$1(Main.scala:131)
	at org.jetbrains.jps.incremental.scala.remote.Main$.handleCommand(Main.scala:138)
	at org.jetbrains.jps.incremental.scala.remote.Main$.serverLogic(Main.scala:115)
	at org.jetbrains.jps.incremental.scala.remote.Main$.nailMain(Main.scala:71)
	at org.jetbrains.jps.incremental.scala.remote.Main.nailMain(Main.scala)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.martiansoftware.nailgun.NGSession.run(NGSession.java:319)
```

dubinsky commented:
> I am getting no issues in Intellij either with Gradle or Intellij building it, are you on the newest version?

This is embarrassing: I **thought** I was, but turns out I was not.
With the latest released Idea and Scala plugin both problems are gone.
Thank you!!!

dubinsky commented:
> As far as the positions go...

When Idea is configured to build directly, positions are reported correctly and messages click through to the correct source positions. When Idea builds using Gradle, positions are reported as character numbers...

dubinsky commented:
@tgodzik Positions like `<1..2..3>` and `[1..2]` 
- come from [dotty.tools.dotc.util.Spans.Span.toString()](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/util/Spans.scala#L135)
- which is called from [dotty.tools.dotc.util.SourcePosition.toString()](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/util/SourcePosition.scala#L81)
- which is called from [dotty.tools.xsbt.PositionBridge.toString()](https://github.com/lampepfl/dotty/blob/master/sbt-bridge/src/dotty/tools/xsbt/PositionBridge.java#L125) (Scala 3 compiler bridge implementation of `xsbti.Position`)
- which is called from [sbt.internal.inc.LoggedReporter.logError()](https://github.com/sbt/zinc/blob/develop/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/LoggedReporter.scala#L158)
- which is called from [sbt.internal.inc.LoggedReporter.display()](https://github.com/sbt/zinc/blob/develop/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/LoggedReporter.scala#L162)
- which is called from [sbt.internal.inc.LoggedReporter.log()](https://github.com/sbt/zinc/blob/develop/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/LoggedReporter.scala#L126).

Before logging the `Problem`, `LoggedReporter` maps its `Position` using [sourcePositionMapper](https://github.com/sbt/zinc/blob/develop/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/LoggedReporter.scala#L105) supplied at the `LoggedReporter`'s creation. Gradle Scala plugin supplies [identity](https://github.com/tgodzik/gradle/blob/add-scala3/subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompiler.java#L139) transformation (resulting in the positions we see); instead, `sourcePositionMapper` should be supplied that maps into an implementation of `xsbti.Position` with the desired `toString()`.

GavinRay97 commented:
There is one tangential issue, just as a heads up for anyone looking to use this with Metals.
I built Tomasz branch, and set up a `gradle` distribution a New Scala + Gradle project.

```gradle
buildscript {
    repositories {
        mavenLocal()
        mavenCentral()
    }

    dependencies {
        classpath "ch.epfl.scala:gradle-bloop_2.12:1.4.8-106-8722ebfd"
    }
}

plugins {
    id "java"
    id "scala"
    id "io.quarkus"
}

// Bloop plugin uses old DSL -- MUST be applied like this
apply plugin: "bloop"

repositories {
    mavenCentral()
    mavenLocal()
}

dependencies {
    // scala3-compiler_3 will pull in corresponding scala3-library_3, which will also pull in corresponding scala-library (Scala 2), 2.13.6 in this case
    implementation "org.scala-lang:scala3-compiler_3:3.0.2-RC1"
}

group = "org.example"
version = "1.0.0-SNAPSHOT"

java {
    sourceCompatibility = JavaVersion.VERSION_16
    targetCompatibility = JavaVersion.VERSION_16
}

tasks.withType(ScalaCompile) {
    scalaCompileOptions.additionalParameters = [
        "-feature",
        "-explain",
        "-Ysafe-init",
        "-Xsemanticdb"
    ]
}

compileJava {
    options.encoding = 'UTF-8'
    options.compilerArgs << '-parameters'
}

compileTestJava {
    options.encoding = 'UTF-8'
}
```

Now the issue is, that this won't work with Metals. It gets detected as a Scala 2.13.6 project, and Scala 3 syntax will error in VS Code 🙁 

You can check the `.bloop/<project-name>.json` file:
```json
    "scala": {
      "organization": "org.scala-lang",
      "name": "scala-compiler",
      "version": "2.13.6",
      "options": [
        "-Xsemanticdb",
        "-Ysafe-init",
        "-deprecation",
        "-explain",
        "-feature",
        "-unchecked"
      ],
```

And `Metals Doctor` will show `2.13.6`

I fixed this by replacing `"version": "2.13.6"` with `"version": "3.0.2-RC1" in both `.bloop/<project-name>.json` and ``.bloop/<project-name>-test.json` and then reloading the window.

But I think this might need a fix in the Gradle plugin.

Anyways, hope this is helpful for anyone that tries to test this in Metals and gets confused/runs into the same thing! 🙂

tgodzik commented:
@dubinsky good catch, I will look into that!

@GavinRay97 ach yes, I was actually checking that out and haven't yet worked on a fix in the Bloop plugin. I will work on it a bit later.

Arthurm1 commented:
@GavinRay97 

For now - adding this to your Gradle build after you apply the plugin should work...
```
bloop {
    stdLibName = "scala3-library_3"
}
```

dubinsky commented:
@tgodzik Positions seem to look more reasonable when appropriate line in `ZincScalaCompiler` is changed from
```java
 new LoggedReporter(100, new SbtLoggerAdapter(), p -> p),
```
to
```java
new LoggedReporter(100, new SbtLoggerAdapter(), PositionImpl::new),
```
where `subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/PositionImpl.java` contains:
```java
package org.gradle.api.internal.tasks.scala;

import xsbti.Position;

import java.io.File;
import java.util.Optional;

public class PositionImpl implements Position {
    private final Position delegate;

    public PositionImpl(final Position delegate) {
        this.delegate = delegate;
    }

    @Override public Optional<Integer> line() { return delegate.line(); }
    @Override public String lineContent() { return delegate.lineContent(); }
    @Override public Optional<Integer> offset() { return delegate.offset(); }
    @Override public Optional<Integer> pointer() { return delegate.pointer(); }
    @Override public Optional<String> pointerSpace() { return delegate.pointerSpace(); }
    @Override public Optional<String> sourcePath() { return delegate.sourcePath(); }
    @Override public Optional<File> sourceFile() { return delegate.sourceFile(); }
    @Override public Optional<Integer> startOffset() { return delegate.startOffset(); }
    @Override public Optional<Integer> endOffset() { return delegate.endOffset(); }
    @Override public Optional<Integer> startLine() { return delegate.startLine(); }
    @Override public Optional<Integer> startColumn() { return delegate.startColumn(); }
    @Override public Optional<Integer> endLine() { return delegate.endLine(); }
    @Override public Optional<Integer> endColumn() { return delegate.endColumn(); }

    @Override public String toString() {
        final Optional<String> sourcePath = this.sourcePath();
        final Optional<Integer> line = this.line();
        final Optional<Integer> column = this.pointer();

        if (sourcePath.isPresent() && line.isPresent() && column.isPresent())
            return sourcePath.get() + ":" + line.get() + ":" + (column.get()+1);
        else if (sourcePath.isPresent() && line.isPresent())
            return sourcePath.get() + ":" + line.get();
        else
            return sourcePath.orElse("");
    }
}
```

This seems to work with both Scala 2 and Scala 3.

dubinsky commented:
@tgodzik In IntelliJ Idea, something adds to `scala-compiler.xml`:
```xml
<parameters>
  <parameter value="-target:jvm-11" />
</parameters>
```
which causes `scalac: bad option '-target:jvm-11' was ignored` when using Scala 3.

I am not sure if Idea's Scala plugin or my configuration needs to be adjusted, but since this does not break anything **and** seems to be unrelated to the Gradle Scala plugin, it is not a concern for your pull request (although it would be nice to get rid of the annoying message...).

tgodzik commented:
@dubinsky Thanks for digging into this! I think the `PositionImpl` is the correct approach, at least with this Zinc version, maybe it will work better in the later one. I will also see if I can reproduce the additional flag being added.

dubinsky commented:
> @dubinsky Thanks for digging into this!

My pleasure ;)

> I think the `PositionImpl` is the correct approach, at least with this Zinc version, maybe it will work better in the later one.

The only change in this area seems to be [this](https://github.com/sbt/zinc/commit/6c877d347363ade53e54cac43ea3a3f3380a5045#diff-dd6958ebc5f904af9157174864ef933e18fd808387ab69b6174623effe3fd186), so it is likely that `PositionImpl` will still be  required with the latest Zinc...

> I will also see if I can reproduce the additional flag being added.

Thanks!

Once you add the position-reporting tweak to your pull request, I think it is time to try and merge it :)

tgodzik commented on subprojects/scala/src/integTest/groovy/org/gradle/scala/compile/Scala3CompileIntegrationTest.groovy:
```diff
@@ -0,0 +1,90 @@
+"""
+        expect:
+        fails 'compileScala'
+        result.assertHasErrorOutput("Person.scala:6:11: Found:    (23 : Int)")
```
@dubinsky Added a test for positions and that seems to work correct. We could try improving the output for sure, but maybe after zinc migration to make sure that we don't have to change too many things.

dubinsky replied:
> Added a test for positions and that seems to work correct

Excellent!

> We could try improving the output for sure, but maybe after zinc migration to make sure that we don't have to change too many things.

I completely agree!

Now, what would it take to get this reviewed/merged/released?

tgodzik replied:
I am trying to get some information on that in the Gradle Slack.

dubinsky replied:
> I am trying to get some information on that in the Gradle Slack.

While we wait, it may make sense to update `subprojects/docs/src/docs/userguide/jvm/scala_plugin.adoc` to mention support for Scala 3 and describe the new library name structure, new major version suffix etc....

tgodzik replied:
Good idea, added! It seems that this is a holiday season in the Gradle team, which means that it might take them a while to get to this.

dubinsky reviewed (approved):
Yay!

tgodzik commented:
I don't seem to be getting `scalac: bad option '-target:jvm-11' was ignored` though I can see it in the `.ipr` file, so maybe the xml you have is leftover? I don't have `scala-compiler.xml` myself. Is it added by a setting maybe?

dubinsky commented:
> I don't have `scala-compiler.xml` myself. Is it added by a setting maybe?

Thanks for looking into this!
I'll figure it out later; the main thing is to get Scala 3 support into Gradle ;)

dubinsky commented on subprojects/docs/src/docs/userguide/jvm/scala_plugin.adoc:
```diff
@@ -129,6 +129,14 @@ include::sample[dir="snippets/scala/quickstart/groovy",files="build.gradle[tags=
+include::sample[dir="snippets/scala/quickstart/groovy",files="build.gradle[tags=scala3-dependency]"]
+include::sample[dir="snippets/scala/quickstart/kotlin",files="build.gradle.kts[tags=scala3-dependency]"]
+====
+
```
Thanks!!!

willisjtc commented:
Any idea when we'll see scala 3 listed here? https://search.maven.org/artifact/org.scala-lang/scala-library

dubinsky commented:
> Any idea when we'll see scala 3 listed here? https://search.maven.org/artifact/org.scala-lang/scala-library

Never. Artifact name changed; see https://search.maven.org/artifact/org.scala-lang/scala3-library_3

willisjtc commented:
> > Any idea when we'll see scala 3 listed here? https://search.maven.org/artifact/org.scala-lang/scala-library
> 
> Never. Artifact name changed; see https://search.maven.org/artifact/org.scala-lang/scala3-library_3

Ooooooo!!!!!

So this is ready to be used now, as a release candidate?

dubinsky commented:
> So this is ready to be used now, as a release candidate?

If by "this" you mean "this pull request" - no; hope is for Scala 3 support to become available in Gradle v7.3...

ljacomet commented on subprojects/docs/src/snippets/scala/quickstart/groovy/build.gradle:
```diff
@@ -19,6 +19,19 @@ dependencies {
 
+
+// tag::scala3-dependency[]
+repositories {
```
This applies to the Groovy and Kotlin ones

tgodzik replied:
The only common dependency is `testImplementation 'junit:junit:4.13'`, so I wasn't sure how best to do it here.

tgodzik replied:
Just having a separate tag fro dependencies would be ok?

ljacomet replied:
I really would duplicate what's in `snippet/scala/quickstart` to `snippet/scala/scala3` and adapt the dependencies section to use Scala 3

tgodzik replied:
I finally understood what you meant! I duplicated the project with Scala 3 dependencies and also changed it to use some Scala 3 specific syntax (braceless)

ljacomet commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/internal/IdeaScalaConfigurer.java:
```diff
@@ -138,6 +138,9 @@ private static ProjectLibrary createScalaSdkLibrary(Project scalaProject, Iterab
         if (runtime != null) {
             FileCollection scalaClasspath = runtime.inferScalaClasspath(files);
             File compilerJar = runtime.findScalaJar(scalaClasspath, "compiler");
+            if (compilerJar == null) {
```
Should there be a notion of priority there?
Like ignore `compiler` if `compiler_3` is present?

tgodzik replied:
There shouldn't be both on the path (and we define it ourselves here), but I could switch it around.

tgodzik replied:
I left it as is for now, so if there is a conflict on the classpath, we choose Scala 2 (in reality this should never happen unless something goes very wrong)

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java:
```diff
@@ -195,7 +231,7 @@ private static File findFile(String prefix, ClassPath classpath) {
         try {
             Properties props = new Properties();
-            props.load(scalaClassLoader.getResourceAsStream("library.properties"));
+            props.load(scalaClassLoader.getResourceAsStream("compiler.properties"));
```
Can you explain this change?
What does it mean for older versions having a `library.properties`?

tgodzik replied:
`library.properties` is contained in `scala_library` and `compiler.properties` is in `scala_compiler` artifact. Both need to be on the classpath and it makes more sense to use the latter. In case of Scala 3 we still use 2.13 standard library, so we can't use `library.properties`

We already do the same thing in Bloop, which uses Zinc underneath https://github.com/scalacenter/bloop/blob/64d8184c812178a3ad949ce38dbba0b8cd4fb53a/backend/src/main/scala/sbt/internal/inc/BloopComponentCompiler.scala#L183

ljacomet reviewed (commented):
Thank you very much for providing this PR and for the community to be responsive in testing it out.
A couple questions, which might come from lack of Scala knowledge, to make sure we do not break existing builds.

I am also sending this through CI to see if it keeps all existing tests happy 😉

ljacomet commented:
@bot-gradle test QF

bot-gradle commented:
<!-- {"replyTargetCommentId":910278798,"teamCityBuildId":"45751183","teamCityBuildHeadRef":"79dd557b2de8f00b5b5961ef99ce36e59e0a3727","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedback build](https://builds.gradle.org/viewQueued.html?itemId=45751183) for you.

ljacomet commented:
@tgodzik Any chance you can rebase this PR on the latest Gradle master?
Looks like CI configuration evolved and running tests needs this branch to have a more recent base.

tgodzik commented:
> @tgodzik Any chance you can rebase this PR on the latest Gradle master?
> Looks like CI configuration evolved and running tests needs this branch to have a more recent base.

Rebased on master and pushed it again!

ljacomet commented:
@bot-gradle test QF

bot-gradle commented:
<!-- {"replyTargetCommentId":910320871,"teamCityBuildId":"45752506","teamCityBuildHeadRef":"b3f5e526079fdd577bff699e48bd5703fd0c78f7","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedback build](https://builds.gradle.org/viewQueued.html?itemId=45752506) for you.

tgodzik commented:
@ljacomet the 12 tests seem to be failing due to https://github.com/scala/bug/issues/10603

Scala 2.11.12 will not work with JDK 9+ - though I am not sure how my change influences this :O

I could update those tests to use 2.12.14 and that should work, but I have no idea why this didn't pop up before.

What do you think?

ljacomet commented:
The test errors seem related to the classloading changes in `ZincScalaCompilerFactory` and the usage of `ScalaCompilerLoader`.
Maybe these changes should only occur when attempting to use Scala 3? And for Scala 2, keep the existing code.

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java:
```diff
@@ -113,43 +129,63 @@ static ZincScalaCompiler getCompiler(CacheRepository cacheRepository, HashedClas
             scalaVersion,
             scalaClassLoader,
-            getClassLoader(DefaultClassPath.of(libraryJar)),
+            scalaLibraryClassLoader,
```
If I undo this change and use `getClassLoader(DefaultClassPath.of(libraryJar), null)` instead and add a null check in `getClassLoader` and similarly pass null as the second param to the call to `getCachedClassLoader` on line 175, then the test passes again.

tgodzik replied:
I can change that back for Scala 2 for sure. However, I was almost 100% sure that scala works with later jdks only since 2.12 🤔

ljacomet replied:
Well I ran tests locally with 11 and CI with 16 and they "work" with the previous setup.

tgodzik replied:
Thanks! I will get back to it tomorrow, not sure why normally it doesn't work with later JDKs, but maybe there is a difference here in Gradle that makes it work, no idea.

I will revert the change for Scala 2 to make the older tests work.

tgodzik replied:
I changed the classloader for Scala 2 to not use any changes made for Scala 3. Not 100% sure what's going on here, but that should be pretty safe to include now.

tgodzik commented:
> The test errors seem related to the classloading changes in `ZincScalaCompilerFactory` and the usage of `ScalaCompilerLoader`.
> Maybe these changes should only occur when attempting to use Scala 3? And for Scala 2, keep the existing code.

I changed the code to use the old classloading mechanism for Scala 2, it should work correctly now. Could you run the tests and see if it helped?

I also changed the snippet to be a separate one to the basic quickstart.

Thanks a lot for the review!

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java:
```diff
@@ -113,43 +130,72 @@ static ZincScalaCompiler getCompiler(CacheRepository cacheRepository, HashedClas
+                }
             }
-            return new URLClassLoader(urls.toArray(new URL[0]));
+            return new URLClassLoader(urls.toArray(new URL[0]), parent);
```
Looking at the class hierarchy, passing `null` here is not the same as calling the single argument constructor - in the later case you implicitly get the system classloader as the parent.

tgodzik replied:
Och! I was sure it is the same, I need to finally download JDK 16 and test it out.

ljacomet replied:
And indeed fixing that call fixes some of the test failures.

ljacomet reviewed (commented):
One more comment around the classloader construction:

ljacomet commented:
@bot-gradle test QF

bot-gradle commented:
<!-- {"replyTargetCommentId":913428493,"teamCityBuildId":"45838758","teamCityBuildHeadRef":"68eb5322c9d63e499af1e505274ecb7741bf83ba","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedback build](https://builds.gradle.org/viewQueued.html?itemId=45838758) for you.

ljacomet commented:
@bot-gradle test QF

bot-gradle commented:
<!-- {"replyTargetCommentId":913475742,"teamCityBuildId":"45841939","teamCityBuildHeadRef":"7ebc7573bd9a81e50db0e5fdee0e24626ecad80d","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedback build](https://builds.gradle.org/viewQueued.html?itemId=45841939) for you.

tgodzik commented:
It works! Thanks a lot for the help @ljacomet ! Should I do something more with the PR?

ljacomet commented:
@bot-gradle test RFM

bot-gradle commented:
<!-- {"replyTargetCommentId":913590139,"teamCityBuildId":"45846015","teamCityBuildHeadRef":"7ebc7573bd9a81e50db0e5fdee0e24626ecad80d","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [ReadyForMerge build](https://builds.gradle.org/viewQueued.html?itemId=45846015) for you.

ljacomet commented:
@tgodzik Thanks a lot for the work.

If you are willing, consider adding an entry to `notes.md` following the template in there to have this feature properly announced in the release notes!

I also launched a longer test suite to check there are no surprises with your changes.

tgodzik commented:
> @tgodzik Thanks a lot for the work.
> 
> If you are willing, consider adding an entry to `notes.md` following the template in there to have this feature properly announced in the release notes!
> 
> I also launched a longer test suite to check there are no surprises with your changes.

Will do in a sec!

tgodzik commented:
@ljacomet I added the additional point to release notes and the tests seemed to all pass :tada:

ljacomet commented:
@bot-gradle test and merge

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=45848092) for you.

bot-gradle commented:
[Pre-tested commit build](https://builds.gradle.org/viewLog.html?buildId=45848092) failed.

ljacomet commented:
@bot-gradle test and merge

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=45852817) for you.

Arthurm1 commented:
You should be able to test this by running...

`./gradlew wrapper --gradle-distribution-url=https://services.gradle.org/distributions-snapshots/gradle-7.3-20210906222431+0000-bin.zip`

Thanks a lot @tgodzik and co for this PR
